src/main/java/com/aws/greengrass/deployment/errorcode/DeploymentErrorCodeUtils.java

```diff
@@ -0,0 +1,213 @@
+import static com.aws.greengrass.deployment.errorcode.DeploymentErrorCode.S3_RESOURCE_NOT_FOUND;
+import static com.aws.greengrass.deployment.errorcode.DeploymentErrorCode.S3_SERVER_ERROR;
+import static com.aws.greengrass.deployment.errorcode.DeploymentErrorCode.SERVER_ERROR;
+import static com.aws.greengrass.deployment.errorcode.DeploymentErrorCode.THROTTLING_ERROR;
+
+public final class DeploymentErrorCodeUtils {
+
+    private static final List<Class<? extends Exception>> NETWORK_OFFLINE_EXCEPTION =
+            Arrays.asList(DeviceConfigurationException.class, SdkClientException.class);
+
+    private static final String COMPONENT_ARN_TEMPLATE = "arn:%s:greengrass:%s:%s:components:%s";
+    private static final String PARTITION_PATTERN_STRING = "(aws(-[a-z]+)*)";
+    private static final String REGION_PATTERN_STRING = "([a-z]{2}(-[a-z]+)+-\\d{1})";
+    private static final String COMPONENT_ACCOUNT_ID_PATTERN_STRING = "(aws|\\d{12})";
+    private static final String COMPONENT_NAME_WITH_OR_WITHOUT_VERSION_PATTERN_STRING =
+            "([a-zA-Z0-9-_\\.]+)" + "(:versions:([^:/\\\\]+))?";
+    private static final String COMPONENT_ARN_PATTERN_STRING =
+            String.format(COMPONENT_ARN_TEMPLATE, PARTITION_PATTERN_STRING, REGION_PATTERN_STRING,
+                    COMPONENT_ACCOUNT_ID_PATTERN_STRING, COMPONENT_NAME_WITH_OR_WITHOUT_VERSION_PATTERN_STRING);
+    private static final Pattern COMPONENT_ARN_PATTERN = Pattern.compile(COMPONENT_ARN_PATTERN_STRING);
+
+
+    private DeploymentErrorCodeUtils() {
+    }
+
+    /**
+     * Walk through exception chain and generate deployment error report.
+     *
+     * @param e exception passed to DeploymentResult
+     * @return error code stack and error types in a pair
+     */
+    public static Pair<List<String>, List<String>> generateErrorReportFromExceptionStack(Throwable e) {
+        // Use a linked hash set to remove duplicates while preserving order
+        Set<DeploymentErrorCode> errorCodeSet =
+                new LinkedHashSet<>(Collections.singletonList(DeploymentErrorCode.DEPLOYMENT_FAILURE));
+        Map<String, DeploymentErrorCode> errorContext = new HashMap<>();
+
+        // iterating through the chain
+        Throwable temp = e;
+        while (temp != null) {
+            translateExceptionToErrorCode(errorCodeSet, temp, errorContext);
+            temp = temp.getCause();
+        }
+
+        List<String> errorStack = errorCodeSet.stream().map(Enum::toString).collect(Collectors.toList());
+        // remove duplicate types
+        List<String> errorTypes = errorCodeSet.stream().map(DeploymentErrorCode::getErrorType).distinct()
+                .filter(type -> !type.equals(DeploymentErrorType.NONE)).map(Enum::toString)
+                .collect(Collectors.toList());
+        return new Pair<>(errorStack, errorTypes);
+    }
+
+    private static void translateExceptionToErrorCode(Set<DeploymentErrorCode> errorCodeSet, Throwable e,
```

Comment:
should you log when it is an unhandled exception type?

Reply:
We synced offline. Since we are always logging the original exception when deployment fails, there is no need to add additional log here.